gnomad_qc/v3/annotations/generate_freq_data.py

```diff
@@ -1,104 +1,260 @@
-        mt.rows().write("gs://gnomad-tmp/gnomad_freq/chr20_1_1000000_freq.ht", overwrite=True)
-    else:
-        mt.rows().write(freq.path, overwrite=args.overwrite)
+        else:
+            logger.info("Computing age histograms for each variant...")
+            mt = mt.annotate_rows(
+                **age_hists_expr(
+                    mt.adj,
+                    mt.GT,
+                    hl.if_else(
+                        hl.is_defined(mt.meta.project_meta.age),
+                        mt.meta.project_meta.age,
+                        mt.meta.project_meta.age_alt,
+                    ),
+                )
+            )
+
+            # Compute callset-wide age histogram global
+            meta_ht = meta.ht()
+            meta_ht = meta_ht.filter(meta_ht.release)
+            mt = mt.annotate_globals(
+                age_distribution=meta_ht.aggregate(
+                    hl.agg.hist(
+                        hl.if_else(
+                            hl.is_defined(meta_ht.project_meta.age),
+                            meta_ht.project_meta.age,
+                            meta_ht.project_meta.age_alt,
+                        ),
+                        30,
+                        80,
+                        10,
+                    )
+                )
+            )
+
+            mt = annotate_freq(
+                mt,
+                sex_expr=mt.meta.sex_imputation.sex_karyotype,
+                pop_expr=mt.meta.population_inference.pop,
+                downsamplings=DOWNSAMPLINGS,
+            )
+            # Remove all loci with raw AC=0
+            mt = mt.filter_rows(mt.freq[1].AC > 0, keep=True)
+
+            logger.info("Calculating InbreedingCoefficient...")
+            # NOTE: This is not the ideal location to calculate this, but added here to avoid another densify
+            mt = mt.annotate_rows(
+                InbreedingCoeff=bi_allelic_site_inbreeding_expr(mt.GT)
+            )
+
+            # Select freq, FAF and popmax
+            faf, faf_meta = faf_expr(
+                mt.freq, mt.freq_meta, mt.locus, POPS_TO_REMOVE_FOR_POPMAX
+            )
+            mt = mt.select_rows(
+                "InbreedingCoeff",
+                "freq",
+                faf=faf,
+                popmax=pop_max_expr(mt.freq, mt.freq_meta, POPS_TO_REMOVE_FOR_POPMAX),
+            )
+            mt = mt.annotate_globals(faf_meta=faf_meta)
+            mt = mt.annotate_globals(faf_index_dict=make_faf_index_dict(mt))
+            mt = mt.annotate_rows(
+                popmax=mt.popmax.annotate(
+                    faf95=mt.faf[
+                        mt.faf_meta.index(
+                            lambda x: x.values() == ["adj", mt.popmax.pop]
+                        )
+                    ].faf95
+                )
+            )
+
+            # Annotate quality metrics histograms, as these also require densifying
+            mt = mt.annotate_rows(**qual_hist_expr(mt.GT, mt.GQ, mt.DP, mt.AD, mt.adj))
```

Comment:
Again this might not actually be any cleaner, was just trying to avoid making the three lists, but what is there now looks good, so feel free to disregard.

```
mt = mt.annotate_rows(qual_hists=qual_hist_expr(mt.GT, mt.GQ, mt.DP, mt.AD, mt.adj))
ht = mt.rows()
ht = ht.annotate(
    qual_hists=hl.Struct(
        **{
            i.replace("_adj", ""): ht.qual_hists[i]
            for i in ht.qual_hists
            if "_adj" in i
        }
    ),
    raw_qual_hists=hl.Struct(
        **{i: ht.qual_hists[i] for i in ht.qual_hists if "_adj" not in i}
    ),
)
```

Reply:
Beautiful -- thanks for this very elegant suggestion!